packages/svelte/src/compiler/phases/2-analyze/css/css-prune.js

```diff
@@ -295,10 +336,87 @@ const regex_backslash_and_following_character = /\\(.)/g;
  * @param {Compiler.Css.StyleSheet} stylesheet
+ * @param {boolean} check_has Whether or not to check the `:has(...)` selectors
  * @returns {boolean}
  */
-function relative_selector_might_apply_to_node(relative_selector, rule, element, stylesheet) {
+function relative_selector_might_apply_to_node(
+	relative_selector,
+	rule,
+	element,
+	stylesheet,
+	check_has
+) {
+	// Sort :has(...) selectors in one bucket and everything else into another
+	const has_selectors = [];
+	const other_selectors = [];
+
 	for (const selector of relative_selector.selectors) {
+		if (selector.type === 'PseudoClassSelector' && selector.name === 'has' && selector.args) {
+			has_selectors.push(selector);
+		} else {
+			other_selectors.push(selector);
+		}
+	}
+
+	// If we're called recursively from a :has(...) selector, we're on the way of checking if the other selectors match.
+	// In that case ignore this check (because we just came from this) to avoid an infinite loop.
+	if (check_has && has_selectors.length > 0) {
+		// :has(...) is special in that it means "look downwards in the CSS tree". Since our matching algorithm goes
+		// upwards and back-to-front, we need to first check the selectors inside :has(...), then check the rest of the
+		// selector in a way that is similar to ancestor matching. In a sense, we're treating `.x:has(.y)` as `.x .y`.
+		for (const has_selector of has_selectors) {
+			const complex_selectors = /** @type {Compiler.Css.SelectorList} */ (has_selector.args)
+				.children;
+			let matched = false;
+
+			for (const complex_selector of complex_selectors) {
+				const selectors = truncate(complex_selector);
+				if (
+					selectors.length === 0 /* is :global(...) */ ||
+					apply_selector(selectors, rule, element, stylesheet, check_has)
+				) {
+					// Treat e.g. `.x:has(.y)` as `.x .y` with the .y part already being matched,
+					// and now looking upwards for the .x part.
+					if (
+						apply_combinator(
+							selectors[0]?.combinator ?? descendant_combinator,
+							selectors[0] ?? [],
+							[relative_selector],
+							rule,
+							element,
+							stylesheet,
+							false
+						)
+					) {
+						complex_selector.metadata.used = true;
+						matched = true;
+					}
+				}
+			}
+
+			if (!matched) {
+				if (relative_selector.metadata.is_global && !relative_selector.metadata.is_global_like) {
+					// Edge case: `:global(.x):has(.y)` where `.x` is global but `.y` doesn't match.
+					// Since `used` is set to `true` for `:global(.x)` in css-analyze beforehand, and
+					// we have no way of knowing if it's safe to set it back to `false`, we'll mark
+					// the inner selector as used and scoped to prevent it from being pruned, which could
+					// result in a invalid CSS output (e.g. `.x:has(/* unused .y */)`). The result
+					// can't match a real element, so the only drawback is the missing prune.
+					// TODO clean this up some day
+					complex_selectors[0].metadata.used = true;
```

Comment:
Maybe put `complex_selectors[0]` as its own better named variable. Why only the first element too?

Reply:
It would just be called `first_complex_selector` which IMO doesn't add any clarity.

As for why only the first: a complex selector is the a and b in `:has(a, b)`. This logic is just there to prevent invalid CSS, which means we only need to mark one of them as used so there's at least one selector in there.

Reply:
I think this caused:

```
    .dropzone:has(+ .dragenter-top),
    :global(.dragenter-bottom + .dropzone) {
        background-color: blue;
        position: relative;
    }
```

to trigger

```
editor:build: warnings when minifying css:
editor:build: ▲ [WARNING] Unexpected ")" [css-syntax-error]
editor:build: 
editor:build:     <stdin>:1230:71:
editor:build:       1230 │ ...) .dropzone:has(/* (unused) + .dragenter-top*/.svelte-12stbyo)*/
editor:build:            ╵   
```

Reply:
Please open an issue with a reproduction

Reply:
Sure: https://github.com/sveltejs/svelte/issues/13680